src/shamir.rs

```diff
@@ -0,0 +1,398 @@
+use rand_core::RngCore;
+use std::iter::repeat_with;
+use std::num::NonZeroU8;
+use thiserror::Error;
+
+/// Shamir secret sharing
+pub struct SecretSharing {
+    /// Threshold
+    k: u8,
+
+    /// Number of shares produced by the `share` method
+    n: u8,
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("Wrong or insecure secret sharing scheme. (expected {k} >= {n} > 1)")]
+    BadSharingScheme { k: u8, n: u8 },
+    #[error("Not enough shares to reconstruct secret: need at least {actual}, got {required}")]
+    NotEnoughShares { required: u8, actual: u8 },
+    #[error("Prime field element inversion failed: {v:?}")]
+    InvertError { v: Box<[u8]> },
+}
+
+/// Represents a single share: (x, f(x)) point
+#[derive(Clone)]
+pub struct Share<F> {
+    x: u8,
+    y: F,
+}
+
+#[derive(Debug, Error)]
+pub enum ShareError {
+    #[error("Can't perform operation on shares as they don't have the same X coordinate: ({lhs} != {rhs})")]
+    MisalignedShares { lhs: u8, rhs: u8 },
+}
+
+impl SecretSharing {
+    /// Constructs a new instance, returning `ShamirError` if input values do not allow
+    /// to construct a valid Shamir secret sharing scheme.
+    ///
+    /// # Errors
+    /// Rejects the following inputs:
+    /// * k == 1
+    /// * k > n
+    pub fn new(k: NonZeroU8, n: NonZeroU8) -> Result<Self, Error> {
+        if k.get() == 1 || k > n {
+            Err(Error::BadSharingScheme {
+                k: k.get(),
+                n: n.get(),
+            })
+        } else {
+            Ok(Self {
+                k: k.get(),
+                n: n.get(),
+            })
+        }
+    }
+
+    /// Minimum number of shares required to reconstruct a secret
+    #[must_use]
+    pub fn threshold(&self) -> u8 {
+        self.k
+    }
+
+    /// Consumes and splits a given secret into `n` shares. only `k` <= `n` shares is required
+    /// to reconstruct it later
+    pub fn split<R: RngCore, F: PrimeField>(&self, secret: F, rng: R) -> Vec<Share<F>> {
+        // generate polynomial of k-1 degree
+        let coefficients = Self::gen_polynomial::<R, F>(rng)
+            .take(usize::from(self.k - 1))
+            .collect::<Vec<_>>();
+
+        let mut shares = Vec::with_capacity(self.n as usize);
+
+        // sample n points by evaluating the polynomial
+        for i in 1..=self.n {
+            let mut y = F::zero();
+            let x = F::from(u64::from(i));
+            for c in coefficients.iter().rev() {
+                y += *c;
+                y *= x;
+            }
+
+            y += secret;
+
+            shares.push(Share { x: i, y });
+        }
+
+        shares
+    }
+
+    /// Reconstructs a secret from a set of size at least k of shares.
+    ///
+    /// # Errors
+    /// Returns an error if there is no enough shares to reconstruct the secret or
+    /// if an error occurred while evaluating the polynomial
+    pub fn reconstruct<F: PrimeField>(&self, shares: &[Share<F>]) -> Result<F, Error> {
```

Comment:
Honestly, I would consider taking in the Lagrange coefficients as an input, since there is really no need to re-calculate them every time.

Reply:
I think it is nice that the caller does not need to know how to compute them. Maybe this implementation is too generic for IPA, I don't see how we can cache them in the case when any `k` out of `n` of shares are provided to reconstruct a secret. I maybe reading https://en.wikipedia.org/wiki/Lagrange_polynomial wrong though - to me it looks like LC depend on specific points that are provided to compute the polynomial: i.e. if I have 3 points (1, 2, 3) and need to reconstruct a linear polynomial, LC would be different if I use points (1,2) or (2,3) for that.


Having said that, in IPA we probably don't need the generic implementation as we seem to always reconstruct from n. I could probably have this interface and leave callers responsible for correctness:

```rust
  pub fn reconstruct<F: PrimeField>(&self, shares: &[Share<F>], lagrange_coeffs: Option<&[F]>) -> Result<(F, Box<[F]>), Error)
```

so the first call would return LC back to the caller and they can reuse it later without knowing how exactly this module computed them. We could even do some rudimentary validation by checking that for any x included in shares, we have LC computed and stored inside `lagrange_coeff`. The only case where I think it won't work is trivial `k = 1`. 

any thoughts on that @benjaminsavage @martinthomson ?

Reply:
so the best interface I could come up with is as follows (ignore the fact that it returns a tuple for a moment, this part can be improved easily)

```rust
pub fn reconstruct<'lcl, F: PrimeField>(&self, shares: &[Share<F>], lagrange: Option<&'lcl [F]>) -> Result<(F, Cow<'lcl, [F]>), Error>
``` 

This function fails to produce the correct result in this case which makes me a little uncomfortable

```rust
    #[test]
    fn wrong_result_if_lagrange_coefficients_provided_are_for_different_input() {
        let (k, n) = (2_u8, 3_u8);
        let rng = StepRng::new(1, 1);
        let secret = Fp::from(5);

        let sharing =
            SecretSharing::new(NonZeroU8::new(k).unwrap(), NonZeroU8::new(n).unwrap()).unwrap();

        let shares = sharing.split(secret, rng);
        // this call returns Lagrange coefficients for k = n 
        let (_, lc) = sharing.reconstruct(&shares, None).unwrap();

        // reconstruct will be able to find all the required coefficients, but fails to compute
        // the correct result because it only uses 2 out of 3 shares (we need different LC in this case)
        let actual = sharing.reconstruct(&shares[0..2], Some(&lc)).unwrap().0;
        assert_ne!(actual, secret)
    }
```


an alternative idea is to always require callers to provide `n` shares to reconstruct a secret, deviating from the most generic implementation but allows for efficient LC caching. I think I am leaning towards that

Reply:
Why not create a container for the LC and have a method on that that does reconstruction?

Reply:
At the present time, I can only envision us reconstructing the secret within the context of secure multiplication. In that, you need all 3 shares because you're reconstructing from a quadratic secret sharing. So caching all 3 Lagrange coefficients seems fine to me.

I agree with @martinthomson that putting the reconstruction stuff into a separate container might make sense. I agree that the caller shouldn't need to know how to compute the coefficients.

Reply:
We discussed it yesterday in person and decided to create a wrapper struct that represents Lagrange polynomial and take it as an argument to `reconstruct` function